### ISSUE TITLE
Reproducing Results from Table 2 for Llama-1B

### ISSUE THREAD

**Issue by neerajgangwar** [InternLM/SIM-CoT#7]
Hi,

Following up on #4, I am having trouble reproducing the results reported in Table 2 of the paper (for Llama-1B). I was able to achieve accuracy comparable to the reported SFT-CoT results after training the model for 25 epochs, so this one is fine.

Could you please confirm the values used for the following arguments for both the Coconut baseline and Coconut with SIM-CoT?

1. `max_latent_stage`
2. `num_epochs`

Thanks!
Neeraj

**Comment by neerajgangwar** [comment]
Thank you for your response. I have follow-up questions based on your response.

You mentioned `max_latent_stage` to be 1. Does this mean there is only one latent stage, as opposed to 3 stages used in the original Coconut paper or 5 in your paper with GPT-2?

Additionally, Stage 1 under the above-mentioned training procedure is the Coconut baseline training. If it achieves 39%, shouldn't Table 2 mention 39% instead of 33.2% for the Coconut baseline (without SIM-CoT)?

**Comment by Wiselnn570** [comment]
> You mentioned `max_latent_stage` to be 1. Does this mean there is only one latent stage, as opposed to 3 stages used in the original Coconut paper or 5 in your paper with GPT-2?

Yes.

> Additionally, Stage 1 under the above-mentioned training procedure is the Coconut baseline training. If it achieves 39%, shouldn't Table 2 mention 39% instead of 33.2% for the Coconut baseline (without SIM-CoT)?

Our first stage is **not Coconut**. Under Coconut’s training strategy (with `c_thought = 2` and `max_latent_stage = 3`), the number of implicit tokens is **gradually increased** via curriculum training, reaching `2 × 3 = 6` total implicit tokens. In contrast, **we use six implicit tokens throughout the entire training process from the very beginning**.

Moreover, Coconut **replaces one explicit reasoning step each time a new group of implicit tokens is appended**. Our method does **not** perform this simultaneous add-and-remove operation. Instead, we **only gradually remove explicit reasoning steps**, while keeping the number of implicit reasoning tokens fixed.

**Comment by neerajgangwar** [comment]
#

> ### Experimental Settings
> * **c_thought**: 6
> * **epochs_per_stage**: 1
> * **max_latent_stage**: 1
> 
> ### Training Procedure
> 1. **Stage 1: llm-only Training**
>    In this stage, all defined implicit tokens participate in training from the beginning, and the explicit reasoning steps are gradually removed. By the end of this stage, all explicit steps are completely removed, and the model performance is approximately **~39**.

@Wiselnn570 How do you remove the explicit reasoning steps _**gradually**_ with the above configuration in the current version of the code? From what I understand, the current code removes all explicit reasoning steps if `scheduled_stage` $$>$$ `max_latent_stage`. Assuming explicit CoT to be $$0^{th}$$ `scheduled_stage`, the first `scheduled_stage` will have 6 latent tokens, followed by all but the first explicit reasoning step. And then the next `scheduled_stage` will remove all explicit reasoning steps. Could you please clarify this and provide the configuration accordingly?

Additionally, in #4, you mentioned that the SIM-CoT is continued from a Coconut checkpoint. Does this mean that "Stage 1: llm-only Training" starts from that checkpoint?

**Comment by Wiselnn570** [comment]
My previous wording may have been a bit unclear. What I intended to convey is that, after SFT, we append all the implicit tokens at once to the model and begin training. Once the training stabilizes, we gradually remove the explicit reasoning steps, while keeping the implicit tokens fixed throughout.

Previously, I implemented this through a quick patch in the data loader, which was removed when preparing the open-source release. I will take time to properly refine and clean up the code accordingly.

**Comment by neerajgangwar** [comment]
Hi @Wiselnn570,

Are you referring to the `progressive_train` flag in the `get_cot_with_explainable_latent_dataset` method of `dataset.py` shown below?
https://github.com/InternLM/SIM-CoT/blob/f3ae9679945baff02ac9b3ad48b6217737cedba1/Coconut/dataset.py#L502

Also, I am confused about the total number of latent tokens used for Llama3-1b. In this thread, you have mentioned the following as the configuration:
```
c_thought: 6
max_latent_stage: 1
```
This would result in 6 latent tokens.

However, on #11, the user asked if `c_thought: 2 epochs_per_stage: 3 max_latent_stage: 5` was the correct configuration for Coconut, and you did not disagree. This would mean Coconut uses 10 latent tokens. Also, the paper mentions 13.2 (I assume 10 latent tokens + 2 tokens for start and end latent + 1.2 tokens for the final answer from No-CoT) as the number of tokens. I am not sure which configuration is correct. Could you please confirm this?

Thank you!
